File: sample/src/main/java/com/lynx/wind/recyclersample/view/FilterActivity.java
package com.lynx.wind.recyclersample.view;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.KeyEvent;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.lynx.wind.recycleradapter.FilterAdapter;
import com.lynx.wind.recyclersample.R;
import com.lynx.wind.recyclersample.holder.SimpleHolder;
import com.lynx.wind.recyclersample.model.User;

import java.util.ArrayList;

public class FilterActivity extends AppCompatActivity {

    private RecyclerView recyclerView;
    private EditText editText;

    private ArrayList<User> data = new ArrayList<>();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_filter);

        setSupportActionBar((Toolbar) findViewById(R.id.toolbar));
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        setupView();
    }

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }

    private void setupView() {
        dummyData();

        recyclerView = findViewById(R.id.list);
        editText = findViewById(R.id.query);

        recyclerView.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false));
        recyclerView.setAdapter(new FilterAdapter<SimpleHolder, User>(SimpleHolder.class, data, R.layout.item_user, R.layout.item_header) {
            @Override
            public void onBind(SimpleHolder holder, User data, int Index) {
                holder.bind(data);
            }

            @Override
            public boolean onFiltering(String comparator, User comparedBy) {
                return comparedBy.getName().toLowerCase().contains(comparator);
            }

            @Override
            public void onFilterResult(int count) {
                Toast.makeText(FilterActivity.this, "Found " +count+ " user's", Toast.LENGTH_SHORT).show();
            }
        });

        editText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {}

            @Override
            public void afterTextChanged(Editable s) {
                ((FilterAdapter)recyclerView.getAdapter())
                        .getFilter()
                        .filter(s.toString());
            }
        });
    }

    private void dummyData() {
        data.add(new User(1, "Alex"));
        data.add(new User(2, "Giant"));
        data.add(new User(3, "Josh"));
        data.add(new User(4, "Sam"));
        data.add(new User(5, "Rich"));
        data.add(new User(6, "Tremor"));
        data.add(new User(7, "Proton"));
        data.add(new User(8, "Sidiq"));
        data.add(new User(9, "Michael"));
        data.add(new User(10, "Leonardo"));
    }
}

File: sample/src/main/java/com/lynx/wind/recyclersample/view/FilterActivity.kt
package com.lynx.wind.recyclersample.view

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.view.inputmethod.EditorInfo
import android.widget.Toast
import com.lynx.wind.recycleradapter.FilterAdapter
import com.lynx.wind.recyclersample.R
import com.lynx.wind.recyclersample.holder.SimpleHolder
import com.lynx.wind.recyclersample.model.User
import kotlinx.android.synthetic.main.activity_filter.*

class FilterActivity : AppCompatActivity() {

    val data = ArrayList<User>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_filter)

        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        setupView()
    }

    override fun onSupportNavigateUp(): Boolean {
        return true.also { onBackPressed() }
    }

    private fun setupView() {
        dummyData()

        list.layoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
        list.adapter = object : FilterAdapter<SimpleHolder, User>(SimpleHolder::class.java, R.layout.item_user, data) {
            override fun onBind(holder: SimpleHolder, data: User, index: Int) {
                holder.bind(data)
            }

            override fun onFiltering(comparator: String, comparedBy: User): Boolean {
                return comparedBy.name.toLowerCase().contains(comparator)
            }

            override fun onFilterResult(count: Int) {
                Toast.makeText(this@FilterActivity, "Found $count user's", Toast.LENGTH_SHORT).show()
            }
        }

        query.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_SEARCH) {
                (list.adapter as FilterAdapter<*, *>).filter.filter(query.text.toString())
            }
            return@setOnEditorActionListener false
        }
    }

    private fun dummyData() {
        data.add(User(1, "Alex"))
        data.add(User(2, "Giant"))
        data.add(User(3, "Josh"))
        data.add(User(4, "Sam"))
        data.add(User(5, "Rich"))
        data.add(User(6, "Tremor"))
        data.add(User(7, "Proton"))
        data.add(User(8, "Sidiq"))
        data.add(User(9, "Michael"))
        data.add(User(10, "Leonardo"))
    }
}